### PR TITLE
hydra: prepare proxy process to host multiple pg and proxy

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Another split from https://github.com/pmodels/mpich/pull/5943.

Refactor to let proxy support potentially multiple pg. This is the last preparation for reusing the proxy process to launch multiple spawns.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/ucx
test:mpich/ch3/most
test:mpich/pmi

File: src/pm/hydra/proxy/pmip_cb.c [hunk 1]
```diff
@@ -458,22 +458,214 @@ static HYD_status pmi_listen_cb(int fd, HYD_event_t events, void *userp)
...
+static HYD_status launch_procs(void);
+static int local_to_global_id(int local_id);
+
+static HYD_status handle_launch_procs(int fd)
+{
+    HYD_status status = HYD_SUCCESS;
+    HYDU_FUNC_ENTER();
+
+    status = procinfo(fd);
+    HYDU_ERR_POP(status, "error parsing process info\n");
```
Comment (raffenet) [inline src/pm/hydra/proxy/pmip_cb.c:570]: Process binding initialization should be here when merge conflicts are fixed.

**Review by raffenet — COMMENTED** [review]
This looks OK to me. Want to make sure the binding fix gets included after conflict resolution.

File: src/pm/hydra/proxy/pmip_cb.c (hunk 1 above)
Comment (hzhou) [reply src/pm/hydra/proxy/pmip_cb.c]: Good reminder! I have rebased the PR.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most
test:mpich/pmi

**Review by raffenet — APPROVED** [review]

**Pull request merged**